### PR TITLE
Cache Prometheus persistent disk name before running tests

### PR DESCRIPTION
**Pull request by jkaniuk**
/cc @krzysied
/cc @mm4tt

### CODE REVIEW

File: clusterloader2/pkg/prometheus/experimental.go [hunk 1]
```diff
@@ -75,6 +80,34 @@ func (pc *PrometheusController) trySnapshotPrometheusDisk() (bool, error) {
...
+	}
+	if pc.provider != "gce" && pc.provider != "gke" && pc.provider != "kubemark" {
+		return fmt.Errorf(
+			"snapshotting Prometheus' disk only available for GCP providers (gce, gke, kubemark), provider is: %s", pc.provider)
+	}
+	// Update cache of Prometheus disk metadata
+	err := wait.Poll(
+		10*time.Second,
+		2*time.Minute,
+		pc.tryRetrievePrometheusDiskMetadata)
```
Comment (mm4tt) [inline clusterloader2/pkg/prometheus/experimental.go]: See the comment below first.

Alternatively (to not have to wait.Polls) you can call tryRetrievePrometheusDiskMetadata from trySnaphostPrometheusDisk and only retry the latter.

File: clusterloader2/pkg/prometheus/experimental.go [hunk 2]
```diff
@@ -75,6 +80,34 @@ func (pc *PrometheusController) trySnapshotPrometheusDisk() (bool, error) {
...
+	if pc.diskMetadata.name == "" || pc.diskMetadata.zone == "" {
+		klog.Warningf("Missing zone or PD name, aborting snapshot")
+		klog.Infof("PD name=%s, zone=%s", pc.diskMetadata.name, pc.diskMetadata.zone)
+		return err
+	}
+	return pc.trySnapshotPrometheusDisk(pc.diskMetadata.name, pc.diskMetadata.zone)
```
[the remark this answers is not in the document]
Comment (jkaniuk) [reply clusterloader2/pkg/prometheus/experimental.go]: It was not retried before, but I agree - that will be safer. Fixed.

File: clusterloader2/pkg/prometheus/experimental.go [hunk 3]
```diff
@@ -75,6 +80,34 @@ func (pc *PrometheusController) trySnapshotPrometheusDisk() (bool, error) {
...
+	pc.diskMetadata.name = pdName
+	pc.diskMetadata.zone = zone
+	return true, nil
+}
+
+func (pc *PrometheusController) snapshotPrometheusDiskIfEnabled() error {
+	if !*shouldSnapshotPrometheusDisk {
+		return nil
+	}
+	if pc.provider != "gce" && pc.provider != "gke" && pc.provider != "kubemark" {
```
[the remark this answers is not in the document]
Comment (jkaniuk) [reply clusterloader2/pkg/prometheus/experimental.go]: done

File: clusterloader2/pkg/prometheus/experimental.go (hunk 1 above)
Comment (jkaniuk) [reply clusterloader2/pkg/prometheus/experimental.go]: done with separate Poll()

File: clusterloader2/pkg/prometheus/experimental.go [hunk 4]
```diff
@@ -75,21 +89,54 @@ func (pc *PrometheusController) trySnapshotPrometheusDisk() (bool, error) {
...
+	pc.diskMetadata.zone = zone
+	return true, nil
+}
+
+func (pc *PrometheusController) snapshotPrometheusDiskIfEnabled() error {
+	if enabled, err := pc.isEnabled(); !enabled {
+		return err
+	}
+	// Update cache of Prometheus disk metadata
+	err := wait.Poll(
```
Comment (mm4tt) [inline clusterloader2/pkg/prometheus/experimental.go]: NIT: You could just call cachePrometheusDiskMetadataIfEnabled here

File: clusterloader2/pkg/prometheus/experimental.go [hunk 5]
```diff
@@ -75,21 +89,54 @@ func (pc *PrometheusController) trySnapshotPrometheusDisk() (bool, error) {
...
 		} else {
-			klog.Warningf("Incorrect disk name %v: %v. Using default name: %v", *prometheusDiskSnapshotName, err, pdName)
+			klog.Warningf("Incorrect disk name %v: %v. Using default name: %v", *prometheusDiskSnapshotName, err, snapshotName)
 		}
 	}
-	klog.Infof("Snapshotting PD '%s' into snapshot '%s' in zone '%s'", pdName, snapshotName, zone)
+	// Snapshot Prometheus disk
+	return wait.Poll(
+		10*time.Second,
+		1*time.Minute,
```
Comment (mm4tt) [inline clusterloader2/pkg/prometheus/experimental.go]: The gcloud compute disks snapshot can easily take O(minutes) for big disks (e.g. prometheus disks in 5K node clusters), so the timeout should be bigger than 1 min.

Thanks for improving this code!

Comment (jkaniuk) [reply clusterloader2/pkg/prometheus/experimental.go]: ok, changed to 10min

File: clusterloader2/pkg/prometheus/experimental.go (hunk 4 above)
Comment (jkaniuk) [reply clusterloader2/pkg/prometheus/experimental.go]: I prefer to leave it as it is. It seems very similar, but error handling is different.

**Pull request merged**